pilot/pkg/networking/core/v1alpha3/route/route.go

```diff
@@ -735,3 +735,76 @@ func getHashPolicy(push *model.PushContext, dst *networking.DestinationWeight) *
 
 	return nil
 }
+
+type envoyRouteType int
+
+const (
+	envoyPath envoyRouteType = iota
+	envoyPrefix
+	envoyRegex
+	envoyCatchAll
+)
+
+// CombineVHostRoutes will concatenate route blocks from two virtual hosts
+func CombineVHostRoutes(first []route.Route, second []route.Route) []route.Route {
```

Comment:
This seems to have a nasty side effect of even reordering match clauses that come from the same resource, where the user has intentionally got it right. Would it be better to try to leave order that the user has explicitly set intact?

Reply:
Yes. We would have to change our docs that routes will be reordered by paths, prefixes and then regexes

Reply:
Shouldn't we prioritize user intent over heuristics? If the user has a specific order in mind (with regex first, to blackhole something, for example), and writes it down in a single resource, I think it should be respected.